api/views/json_checker.py

```diff
@@ -0,0 +1,95 @@
+        self.path = path
+
+
+class IncorrectShape(JsonCheckerException):
+    def __init__(self, correct: str, path: str):
+        self.correct = correct
+        self.path = path
+
+
+type_map: dict[type, str] = {dict: 'object', list: 'array', str: 'string', int: 'number (int)', type(None): 'null'}
+
+
+def check(shape, data):
+    """Check the data and raise an exception in case of a mismatch."""
+    match shape:
+        case dict():
+            if not isinstance(data, dict):
+                raise IncorrectShape(type_map[dict], '')
+            for key, component in shape.items():
+                if key not in data:
+                    raise MissingField(f'["{key}"]')
+                try:
+                    check(component, data[key])
+                except MissingField as m:
+                    raise MissingField(f'["{key}"]{m.path}')
+                except IncorrectShape as e:
+                    raise IncorrectShape(e.correct, f'["{key}"]{e.path}')
+        case [element_shape]:
+            if not isinstance(data, list):
+                raise IncorrectShape(type_map[list], '')
+            for i, data_element in enumerate(data):
+                try:
+                    check(element_shape, data_element)
+                except MissingField as m:
+                    raise MissingField(f'[{i}]{m.path}')
+                except IncorrectShape as e:
+                    raise IncorrectShape(e.correct, f'[{i}]{e.path}')
+        case Union(members=members):
+            exceptions = []
+            for member in members:
+                try:
+                    check(member, data)
+                    break
+                except JsonCheckerException as e:
+                    exceptions.append(e)
+            else:
+                error_strings = []
+                for e in exceptions:
+                    match e:
+                        case MissingField(path=path):
+                            error_strings.append(f'missing {path}')
+                        case IncorrectShape(correct=correct, path=path):
+                            error_strings.append(f'{correct} at {path}')
+                raise IncorrectShape(f'one of ({" | ".join(error_strings)})', '')
+
+        case type():
+            if not isinstance(data, shape):
+                raise IncorrectShape(type_map[shape], '')
+    return True
+
+
+def check_error(shape, data) -> None | str:
+    """Check the data and return a formatted error string in case of a mismatch."""
+    try:
+        check(shape, data)
+    except MissingField as e:
+        return f'Missing field at data{e.path}'
+    except IncorrectShape as e:
+        return f'Incorrect shape at data{e.path}. Expected {e.correct}.'
+
+
+def abort_with_error(error: None | str):
+    """Check the data and abort the current request with code 400 and a formatted error string in the description."""
+    if error is None:
+        return
+    abort(400, description=error)
```

Comment:
I initially thought it might be useful outside of view functions, but I guess view functions are the only place where we need to validate data coming from the outside.